File: python/problem014.py
#!/usr/bin/python

"""Problem 14: Longest Collatz sequence"""

def main():

    def get_chain_length(n, cache={2: 2}):
        if not n in cache:
            cache[n] = 2 + get_chain_length((3*n+1)>>1) if n&1 else 1 + get_chain_length(n>>1)
        return cache[n]

    limit = 1000000
    longest = num = 0

    # Look only at the second half of the numbers in the range
    # For every number n ≤ limit//2, 2n is also in the range
    for i in range(limit//2, limit):
        length = get_chain_length(i)

        if length > longest:
            longest = length
            num = i

    return num

if __name__ == "__main__":
    print(main())

File: go/problem014.go
package main

import "fmt"

var cache = map[int]int{2: 2}

func get_chain_len(n int) int {

	if cache[n] == 0 {
		if n%2 == 0 {
			cache[n] = 1 + get_chain_len(n/2)
		} else {
			cache[n] = 2 + get_chain_len((3*n+1)/2)
		}
	}
	return cache[n]
}

func main() {
	const lim = 1000000
	longest := 0
	num := 0

	for i := lim / 2; i < lim; i++ {
		length := get_chain_len(i)
		if length > longest {
			longest = length
			num = i
		}
	}
	fmt.Printf("%d\n", num)
}
